File: 07_stacks_and_queues/stone_wall.py
def solution(H):
  stack = [0]
  cnt = 0
  for i in H:
    while i < stack[-1]:
      stack.pop()
      cnt += 1
    if i > stack[-1]:
      stack.append(i)
  cnt += len(stack) - 1
  return cnt

File: 07_stacks_and_queues/stone_wall.cpp
#include "stack"
#include "vector"

using namespace std;

int solution(vector<int> &h) {
  int ret = 0;
  int n = h.size();
  stack<int> stk;
  stk.push(0);
  for (int i = 0; i < n; i++) {
    while (stk.top() > h[i]) {
      stk.pop();
      ret++;
    }
    if (stk.top() < h[i]) {
      stk.push(h[i]);
    }
  }
  ret += stk.size() - 1;
  return ret;
}
